application/comit_node/src/libp2p_bam/behaviour.rs

```diff
@@ -0,0 +1,184 @@
+        self.addresses
+            .get(peer_id)
+            .cloned()
+            .unwrap_or_else(Vec::new)
+    }
+
+    fn inject_connected(&mut self, peer_id: PeerId, endpoint: ConnectedPoint) {
+        log::debug!(target: "bam", "connected to {} at {:?}", peer_id, endpoint);
+
+        let address = match endpoint {
+            ConnectedPoint::Dialer { address } => address,
+            ConnectedPoint::Listener { send_back_addr, .. } => send_back_addr,
+        };
+
+        match self.addresses.entry(peer_id) {
+            Entry::Occupied(mut entry) => {
+                let addresses = entry.get_mut();
+                addresses.push(address)
+            }
+            Entry::Vacant(entry) => {
+                let addresses = vec![address];
+                entry.insert(addresses);
+            }
+        }
+
+        if let Some(task) = &self.current_task {
+            task.notify()
+        }
+    }
+
+    fn inject_disconnected(&mut self, peer_id: &PeerId, endpoint: ConnectedPoint) {
+        log::debug!(target: "bam", "disconnected from {} at {:?}", peer_id, endpoint);
+
+        if let Some(task) = &self.current_task {
+            task.notify()
+        }
+    }
+
+    fn inject_node_event(&mut self, _: PeerId, event: InnerEvent) {
+        match event {
+            InnerEvent::IncomingRequest(pending_incoming_request) => {
+                self.events.push_back(NetworkBehaviourAction::GenerateEvent(
+                    pending_incoming_request,
+                ));
+
+                if let Some(task) = &self.current_task {
+                    task.notify()
+                }
+            }
+            InnerEvent::IncomingResponse(PendingIncomingResponse { response, channel }) => {
+                let _ = channel.send(response);
+            }
+            InnerEvent::BadIncomingRequest(AutomaticallyGeneratedErrorResponse {
+                response,
+                channel,
+            }) => {
+                let _ = channel.send(response);
+            }
+            InnerEvent::Error => {}
+        }
+    }
+
+    fn poll(
+        &mut self,
+        _params: &mut PollParameters<'_>,
+    ) -> Async<NetworkBehaviourAction<PendingOutgoingRequest, PendingIncomingRequest>> {
+        match self.events.pop_front() {
+            Some(event) => {
+                log::debug!(target: "bam", "emitting {:?}", event);
+
+                if let NetworkBehaviourAction::SendEvent { peer_id, .. } = &event {
+                    if !self.addresses.contains_key(peer_id) {
+                        log::info!(
+                            target: "bam",
+                            "not yet connected to {}, cannot send message",
+                            peer_id.clone()
+                        );
+
+                        self.events.push_back(event);
+
+                        self.current_task = Some(task::current());
+                        return Async::NotReady;
+                    }
+                }
```

Comment:
This returning `NotReady` here is unsound IMO.
You have events in the queue that may be ready. You should emit them.

Doing what I suggested above would fix this.
Also if you do that, you can change `self.events` into a `mpsc::unbounded` and just poll that here. This would remove all the error prone storing of `task::current`.